store/src/main/java/org/apache/rocketmq/store/queue/SparseConsumeQueue.java

```diff
@@ -0,0 +1,333 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.rocketmq.store.queue;
+
+import org.apache.rocketmq.common.UtilAll;
+import org.apache.rocketmq.store.MessageStore;
+import org.apache.rocketmq.store.SelectMappedBufferResult;
+import org.apache.rocketmq.store.logfile.MappedFile;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+public class SparseConsumeQueue extends BatchConsumeQueue {
+
+    public SparseConsumeQueue(
+        final String topic,
+        final int queueId,
+        final String storePath,
+        final int mappedFileSize,
+        final MessageStore defaultMessageStore) {
+        super(topic, queueId, storePath, mappedFileSize, defaultMessageStore);
+    }
+
+    public SparseConsumeQueue(
+        final String topic,
+        final int queueId,
+        final String storePath,
+        final int mappedFileSize,
+        final MessageStore defaultMessageStore,
+        final String subfolder) {
+        super(topic, queueId, storePath, mappedFileSize, defaultMessageStore, subfolder);
+    }
+
+    @Override
+    public void recover() {
+        MappedFile lastMappedFile = this.mappedFileQueue.getLastMappedFile();
+        if (lastMappedFile == null) {
+            return;
+        }
```

Comment:
Do I need to add ```synchronized``` or ```lock``` to ensure thread safety when restoring a ByteBuffer from native?

Reply:
the method `recover` invoked by only one thread, I think there is no need to add `synchronized` or `lock`